File: leetcode-cn/14.最长公共前缀.py
#
# @lc app=leetcode.cn id=14 lang=python3
#
# [14] 最长公共前缀
#

# @lc code=start
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        min_len = 9999
        for s in strs:
            if len(s) < min_len:
                min_len = len(s)

        ans = []
        i = 0

        while i < min_len:
            c = ''
            equal = True
            for s in strs:
                if c == '':
                    c = s[i]
                elif s[i] != c:
                    equal = False
                    break

            if not equal:
                break
            ans.append(c)
            i += 1

        return ''.join(ans)


# @lc code=end

File: leetcode-cn/C++/14.最长公共前缀.cpp
/*
 * @lc app=leetcode.cn id=14 lang=cpp
 *
 * [14] 最长公共前缀
 */

// @lc code=start
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
    string help(const string str1, const string str2)
    {
        int i = 0;
        for (; i < str1.size() && i < str2.size(); i++) {
            if (str1[i] != str2[i]) {
                break;
            }
        }
        return str1.substr(0, i);
    }

    string longestCommonPrefix(vector<string>& strs)
    {
        if (strs.size() == 0) {
            return "";
        }

        string prefix = strs[0];
        for (int i = 1; i < strs.size(); i++) {
            prefix = help(prefix, strs[i]);
            if (prefix.size() == 0) {
                break;
            }
        }
        return prefix;
    }
};
// @lc code=end
